pkg/auth/providers/azure/config.go

```diff
@@ -19,30 +18,39 @@ const (
 	chinaAzureMSLoginEndpoint = "https://login.partner.microsoftonline.cn/"
 )
 
+// GraphEndpointMigratedAnnotation is the main piece of data based on which Rancher decides to use either the
+// deprecated authentication flow via Azure AD Graph or the new one via Microsoft Graph.
+// If the annotation is missing on the Auth Config object, or is present with a value of anything other than "true",
+// then Rancher uses the old, deprecated flow. If the annotation is present and set to "true", Rancher uses the new flow.
+const GraphEndpointMigratedAnnotation = "auth.cattle.io/azuread-endpoint-migrated"
+
 func authProviderEnabled(config *v32.AzureADConfig) bool {
 	return config.Enabled && config.GraphEndpoint != ""
 }
 
-func graphEndpointDeprecated(endpoint string) bool {
-	deprecatedEndpoints := []string{globalAzureADGraphEndpoint, chinaAzureADGraphEndpoint}
-	return slice.ContainsString(deprecatedEndpoints, endpoint)
+func isConfigDeprecated(cfg *v32.AzureADConfig) bool {
+	if !cfg.Enabled {
+		return false
+	}
+	v, ok := cfg.ObjectMeta.Annotations[GraphEndpointMigratedAnnotation]
+	return !ok || v != "true"
 }
 
-func updateAzureADConfig(c *v32.AzureADConfig) {
+func updateAzureADEndpoints(c *v32.AzureADConfig) {
 	if isConfigForChina(c) {
-		updateConfigForChina(c)
+		updateEndpointsForChina(c)
 	} else {
-		updateConfigForGlobal(c)
+		updateEndpointsForGlobal(c)
 	}
 }
 
 func isConfigForChina(c *v32.AzureADConfig) bool {
 	return strings.HasSuffix(c.GraphEndpoint, ".cn") || strings.HasSuffix(c.GraphEndpoint, ".cn/")
 }
 
-func updateConfigForGlobal(c *v32.AzureADConfig) {
+func updateEndpointsForGlobal(c *v32.AzureADConfig) {
 	if c.GraphEndpoint != globalAzureADGraphEndpoint {
-		logrus.Infoln("Refusing to upgrade because the Graph Endpoint is not deprecated.")
+		logrus.Infof("Refusing to upgrade because the Graph Endpoint %s is not deprecated.", c.GraphEndpoint)
```

Comment:
Docs:

One note here: We will want to make sure that we include docs for users of custom endpoints so that they understand that mixing custom/standard endpoints are not supported (as well as what steps are required to upgrade a custom endpoint).

Reply:
Yes, we will say that we're not responsible for any custom invalid endpoints.